src/event_sources/ble_esp32.h

```diff
@@ -95,13 +98,99 @@ class GATTResource : public BLEResource {
   }
 
  private:
-  uint16 _handle;
+  uint16 _handle = kInvalidHandle;
   Mutex* _mutex;
   uint32 _result = 0;
   uint32 _error = 0;
   struct os_mbuf* _mbuf = null;
 };
 
+
+class BLEServerServiceResource;
+typedef LinkedList<BLEServerServiceResource> BLEServerServiceList;
+
+class BLEServerCharacteristicResource;
+typedef LinkedList<BLEServerCharacteristicResource> BLEServerCharacteristicList;
+
+class BLEServerCharacteristicResource: public Resource, public BLEServerCharacteristicList::Element{
+ public:
+  TAG(BLEServerCharacteristicResource);
+  BLEServerCharacteristicResource(ResourceGroup* resource_group, BLEServerServiceResource* service,
+                                  ble_uuid_any_t uuid, int type, struct os_mbuf* value):
+      Resource(resource_group),
+      _service(service),
+      _uuid(uuid),
+      _type(type),
+      _mbuf_to_send(value),
+      _nimble_value_handle(0),
+      _mbuf_received(null),
+      _indicate(false),
+      _notify(false),
+      _conn_handle(0),
+      _mutex(OS::allocate_mutex(3, "")) {}
+
+  ble_uuid_any_t uuid() const { return _uuid; }
+  ble_uuid_t* uuid_p() { return &_uuid.u; }
+  int type() const { return _type; }
+  struct os_mbuf* mbuf_to_send() { Locker locker(_mutex); return _mbuf_to_send; };
```

Comment:
Why is this struct necessary?

Reply:
I would take the methods that are accessed from other threads and group them. Looks like it is the mbuf operations. I wouldn't have any uses of Locker in inline method definitions so move the bodies to the .cc file.